### PR TITLE
[IOTDB-434] Refactor SeriesReaderByTimestamp

### PR DESCRIPTION
**Pull request by liutaohua**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java [hunk 1]
```diff
@@ -75,12 +75,13 @@ public GroupByWithValueFilterDataSet(long queryId, GroupByPlan groupByPlan) {
    * init reader and aggregate function.
    */
   private void initGroupBy(QueryContext context, GroupByPlan groupByPlan)
-      throws StorageEngineException, IOException, PathException {
+      throws StorageEngineException, IOException {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByWithValueFilterDataSet.java]:
```suggestion
      throws StorageEngineException {
```

File: server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  public SeriesDataReaderByTimestamp(Path seriesPath, TSDataType dataType, QueryContext context,
+      QueryDataSource dataSource) {
+    super(seriesPath, dataType, context, dataSource.getSeqResources(),
+        dataSource.getUnseqResources());
+  }
+
+  @Override
+  public Object getValueInTimestamp(long timestamp) throws IOException {
+    this.filter = TimeFilter.gtEq(timestamp);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java]: we may need to set the value in TimeEq filter, construct this object each time is costly.

File: server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+      if (!hasNext(timestamp)) {
+        return null;
+      }
+    }
+
+    return batchData.getValueInTimestamp(timestamp);
+  }
+
+  @Override
+  public boolean hasNext() throws IOException {
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java]: could this method be removed from the IReaderByTimestamp?

File: server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      }
+    }
+
+    return batchData.getValueInTimestamp(timestamp);
+  }
+
+  private boolean hasNext(long timestamp) throws IOException {
+    while (super.hasNextChunk()) {
+      Statistics statistics = firstChunkMetaData.getStatistics();
+      if (statistics.getEndTime() < timestamp) {
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java]: You're absolutely right, I need to refactor part of `AbstractReader` to complete the implementation of this method

File: server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        }
+        if (batchData.getTimeByIndex(batchData.length() - 1) >= timestamp) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  private Statistics currentPageStatistics() throws IOException {
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java]: I think should be ` extends AbstractReader`  directly, `currentPageStatistics` is not the`SeriesDataReaderWithoutValueFilter` method

File: server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+
+  @Override
+  protected boolean satisfyFilter(Statistics statistics) {
+    return true;
+  }
+
+  @Override
+  public boolean hasNext() throws IOException {
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java]: Currently, `ExternalSortJob` is also using `IReaderByTimestamp`, I'm not going to fix it in this pr

File: server/src/main/java/org/apache/iotdb/db/query/reader/resourceRelated/SeqResourceReaderByTimestamp.java [hunk 7]
```diff
@@ -1,190 +1,189 @@
...
+// * <code>IReaderByTimestamp</code> for the TsFiles.
+// * <p>
+// * Notes: 1) The list of sequence TsFiles is in strict chronological order. 2) The data in a
+// * sequence TsFile is also organized in chronological order. 3) A sequence TsFile can be either
+// * sealed or unsealed. 4) An unsealed sequence TsFile consists of two parts of data in chronological
+// * order: data that has been flushed to disk and data in the flushing memtable list.
+// * <p>
+// */
+//
+//public class SeqResourceReaderByTimestamp implements IReaderByTimestamp {
```
[the remark this answers is not in the document]
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/resourceRelated/SeqResourceReaderByTimestamp.java]: done

File: server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java (hunk 2 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java]: no , I think `gtEq` should be used here

File: server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java (hunk 3 above)
Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/seriesRelated/SeriesDataReaderByTimestamp.java]: I'm not going to fix it in this pr

File: server/src/main/java/org/apache/iotdb/db/query/reader/chunkRelated/MemChunkReaderByTimestamp.java [hunk 8]
```diff
@@ -19,31 +19,25 @@
...
   private boolean hasCachedTimeValuePair;
   private TimeValuePair cachedTimeValuePair;
 
-  public MemChunkReaderByTimestamp(List<ReadOnlyMemChunk> readableChunk) throws IOException {
+  public MemChunkReaderByTimestamp(ReadOnlyMemChunk memChunk) throws IOException {
     timeValuePairIterator = new PriorityMergeReader();
-    for (ReadOnlyMemChunk memChunk : readableChunk) {
-      timeValuePairIterator.addReader(memChunk.getIterator(), memChunk.getVersion());
-    }
+    timeValuePairIterator.addReader(memChunk.getIterator(), memChunk.getVersion());
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/reader/chunkRelated/MemChunkReaderByTimestamp.java]: no need to use a  prioritemergereader

Comment (liutaohua) [reply server/src/main/java/org/apache/iotdb/db/query/reader/chunkRelated/MemChunkReaderByTimestamp.java]: done

**Pull request merged**